src/programs/city/defense.js

```diff
@@ -55,27 +68,36 @@ class CityDefense extends kernel.process {
   }
 
   fireTowers (towers, hostiles) {
-    if (hostiles.length > 0) {
-      // for now, just shoot closest for each tower
-      let tower
-      for (tower of towers) {
+    const attackFunc = (attackTarget) => {
+      for (let tower of towers) {
         if (tower.energy < TOWER_ENERGY_COST) {
           continue
         }
-        const closest = _.min(hostiles, c => c.pos.getRangeTo(tower.pos))
-        tower.attack(closest)
+
+        tower.attack(attackTarget)
       }
-      return
     }
 
     const healFunc = (healTarget) => {
-      // TODO: clever calculations to avoid overheals
-      let tower
-      for (tower of towers) {
+      let damage = healTarget.hitsMax - healTarget.hits
+
+      for (let tower of towers) {
+        if (damage <= 0) {
+          break
+        }
+
+        const distance = tower.pos.getRangeTo(healTarget.pos)
+        damage -= global.TOWER_HEAL_EFFECT[distance]
```

Comment:
might break if the target, for whatever reason, is in another room. I'd suggest clamping the range to the max room size